pkg/common/watchdog.go

```diff
@@ -68,3 +68,65 @@ func StartWatchdog(c WatchdogConfig) *WatchdogChannel {
 
 	return &WatchdogChannel{incoming, sync.Mutex{}, false}
 }
+
+type Watchdog struct {
+	// Timeout after which `OnTimeout` is called.
+	timeout time.Duration
+	// A closure that is called once `Timeout` is reached.
+	onTimeout   func()
+	notify      chan struct{}
+	closeSignal chan struct{}
+}
+
+func NewWatchdog(timeout time.Duration, onTimeout func()) *Watchdog {
+	return &Watchdog{timeout: timeout, onTimeout: onTimeout, notify: make(chan struct{}), closeSignal: make(chan struct{})}
+}
```

Comment:
Is there a use case for starting a watchdog that does not get started?

I think the previous API surface of the `WatchdogChannel` was somewhat more resilient in this case, i.e. the watchdog is started once created and could be notified and closed after that.

Whereas the new API has an issue that calling `Notify()` before `Start()` would **hang the caller forever**. Also, it allows closing the `Watchdog` that has not been started in the first place.

I think if we want a drop-in replacement, we need to preserve the API surface.

Another concern that I see is that the `notify` channel is unbounded, this is different from our implementation in `main` that uses buffered channels (since we generally don't want to block the caller, the caller merely informs the watchdog so that it continues to run and not call the `OnTimeout`, i.e. one can treat `Watchdog` as a channel that provides a unidirectional heartbeat to the caller, making sure that if no heartbeat messages are received for `timeout`, some action `onTimeout()` is executed).

Also `Start()` can be called multiple times which would make the behavior of the watchdog a bit unpredictable due to separate goroutines racing to receive the `Notify()` message, where only would get it.

Reply:
About the general remarks:

1. I agree, i would write more comments
2. I agree, one implementation has to leave :-)
3. and 4. I will split the PR into two topics. I will rebase them and note your other comments. I will try to include the ci tests in the PR with pkg structure changes.

Reply:
Cool, thanks Enrico!

Reply:
> I would differentiate a little bit about the unbounded channel part. I'm not that happy with my implementation, because this implementation doesn't make the source code much better. Of course, we can use a buffer for the notify channel, but its remains still blocking. Because, when the buffer is reached then sending is blocked again.
>
> This is not a theoretical view of that. 'Notify()' is called in Subscription.WriteRTP(). This method is called whenever a RTP packet arrives. If I say something wrong please correct me, that can be 400 packets per second.

I totally agree with your reasoning. Theoretically, it can and will reach the buffer size at some point (and practically will reach it, though in our case that would mean a bug on our side). The constant that is defined in a `common` package was meant to set the size of the channel to "large enough" that it sort of "never blocks", or if I can re-formulate it: "if it ever reaches this size and blocks then there is a bug somewhere in our code".

Do you know if there is a better way to create a channel that does not have a limited buffer size (i.e. the buffer of which is limited to the size of the RAM, like unbounded channels in Rust)? Alternatively, the only way to ensure that it never blocks (even in theory) would be to do something like `go func() { ch<- struct{}{}}()` or something like this, but it feels very inefficient to spawn a goroutine just for the sending procedure (I know that goroutines are lighter than actual threads, but are they that light to actually be used like this given that we would be doing it hundreds times per second).

We could of course try to use the `select/default` trick though.

Reply:
I once added a benchmark:
```shell
BenchmarkWatchdog_Notify-4               2801523               418.1 ns/op
BenchmarkWatchdogChannel_Notify-4        4755174               243.7 ns/op
```
An RTP packet arrives every 20 ms, there is theoretically enough time :-) Now I think, both implementations are fast enough. 

Unfortunately, my implementation is twice as slow as the original implementation. :-( So we should stay on the old one.

To your question:
I think go routines are fast enough.
But I cannot answer your question in general. I think the blocking behavior of go channels is explicitly intended. I think channels are there to transfer ownership of data from one go routine to the next. And for this process, the sender has to wait until the receiver is ready.
When go routines are started I always ask myself from which go routine is it getting data and to which go routine is data send.